packages/start/config/server-runtime.jsx

```diff
@@ -13,55 +13,100 @@ import {
 } from 'seroval-plugins/web';
 import { createIslandReference } from "../server/islands";
 
-async function deserializeStream(id, response) {
-  if (!response.body) {
-    throw new Error("missing body");
+class ChunkReader {
+  constructor(stream) {
+    this.reader = stream.getReader();
+    this.buffer = '';
+    this.done = false;
   }
-  const reader = response.body.getReader();
-
-  async function pop() {
-    const result = await reader.read();
-    if (!result.done) {
-      const serialized = new TextDecoder().decode(result.value);
-      const splits = serialized.split("\n");
-      for (const split of splits) {
-        if (split !== "") {
-          deserialize(split);
-        }
-      }
-      await pop();
+
+  async readChunk() {
+    // if there's no chunk, read again
+    const chunk = await this.reader.read();
+    if (!chunk.done) {
+      // repopulate the buffer
+      this.buffer += new TextDecoder().decode(chunk.value);
+    } else {
+      this.done = true;
     }
   }
 
-  let serialized = ""
-  while (true) {
-    const line = await reader.read();
-    if (line.done) break;
-    serialized += new TextDecoder().decode(line.value);
+  async nextValue() {
+    // Check if the buffer is empty
+    if (this.buffer === '') {
+      // if we are already one...
+      if (this.done) {
+        return {
+          done: true,
+          value: undefined,
+        };
+      }
+      // Otherwise, read a new chunk
+      await this.readChunk();
+    }
+    // Get the first valid seroval chunk
+    const [first, ...rest] = this.buffer.split('\n');
+    // Deserialize the seroval chunk
+    const result = {
+      done: false,
+      value: deserialize(first),
+    };
+    // if it succeeds, remove the first valid chunk
+    // from the buffer
+    this.buffer = rest.join('\n');
```

Comment:
`Array.shift` instead of split/destructure/join?

Reply:
I still needed to re-join it since buffer is a string.